### PR TITLE
validate ikey

### PR DESCRIPTION
**Pull request by TimothyMothra**
#1260 

As we discussed, InstrumentationKey will be checked for invalid characters when set.

### CODE REVIEW

File: BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/InstrumentationKeyValidation.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+            if (instrumentationKey == null)
+            {
+                throw new ArgumentNullException(nameof(instrumentationKey));
+            }
+            else if (string.IsNullOrWhiteSpace(instrumentationKey))
+            {
+                throw new ArgumentException("Instrumentation Key can not be empty.");
+            }
+
+            Regex allowedCharacters = new Regex("^[a-zA-Z0-9-:]*$");
```
Comment (Dmitry-Matveev) [inline BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/InstrumentationKeyValidation.cs]: This excludes Unicode alphanumeric character and only includes English alphanumeric - do we expect IKey string to be non-English in some cases? If yes, we can change the Regex to account for that.

**Comment by cijothomas** [comment]
@TimothyMothra Can you describe the validation behavior if ikey is configured via other options? eg: environment variable? And asp.net core offers additional places for provide instrumentation key. (aioptions)
How about ikey provided inside connectionstrings?

**Comment by TimothyMothra** [comment]
@cijothomas 
Environment Variable and ConnectionString are both used to set the telemetryconfiguration.instrumentationkey, so they are covered by these same validation rules.
IIRC aioptions also maps to the telemetryconfiguration.

**Comment by cijothomas** [comment]
> @cijothomas
> Environment Variable and ConnectionString are both used to set the telemetryconfiguration.instrumentationkey, so they are covered by these same validation rules.
> IIRC aioptions also maps to the telemetryconfiguration.

If you can do a unit test asserting these, it'd be great.

**Comment by TimothyMothra** [comment]
Abandoning this, we are going in a different direction.

**Pull request closed without merging**